langchain4j-mcp/src/main/java/dev/langchain4j/mcp/McpToolProvider.java

```diff
@@ -0,0 +1,52 @@
+package dev.langchain4j.mcp;
+
+import dev.langchain4j.agent.tool.ToolSpecification;
+import dev.langchain4j.mcp.client.McpClient;
+import dev.langchain4j.service.tool.ToolProvider;
+import dev.langchain4j.service.tool.ToolProviderRequest;
+import dev.langchain4j.service.tool.ToolProviderResult;
+import java.util.List;
+
+/**
+ * A tool provider backed by one or more MCP clients
+ */
+public class McpToolProvider implements ToolProvider {
+
+    private final List<McpClient> mcpClients;
+
+    private McpToolProvider(Builder builder) {
+        this.mcpClients = builder.mcpClients;
+    }
+
+    @Override
+    public ToolProviderResult provideTools(final ToolProviderRequest request) {
+        ToolProviderResult.Builder builder = ToolProviderResult.builder();
+        for (McpClient mcpClient : mcpClients) {
+            List<ToolSpecification> toolSpecifications = mcpClient.listTools();
+            for (ToolSpecification toolSpecification : toolSpecifications) {
+                builder.add(
+                        toolSpecification, // TODO: how do I handle memory here?
```

Comment:
Why is memory relevant?

`memoryId` is currently there to distinguish between users. This is not ideal though, and I want to change it to accepts a map of attributes instead of a single memory id. This way user will be able to pass some parameters to the tool. But this is probably not relevant if MCP server does not distinguish between end users?

Reply:
Yeah I wasn't sure about it so I put a TODO there, but probably we can ignore the memory